Refactor Bpfd-Agent and write Bpfd-Operator (#212)
redhat-et/bpfd: bpfd-operator/README.md

# bpfd-operator

The Bpfd-Operator repository exists in order to deploy and manage bpfd within a kubernetes cluster.

![bpfd on K8s](./bpfd-on-k8s.jpg)

## Description

This repository houses two main processes, the `bpfd-agent` and the `bpfd-operator` along with CRD api definitions for `EbpfProgram` and `EbpfProgramConfig` Objects. In the below sections we'll dive a bit deeper into the API and functionality of both processes.

But first try it out!

## Getting started

This operator was build utilizing some great tooling provided by the [operator-sdk library](https://sdk.operatorframework.io/). A great first step in understanding some
of the fucntionality can be to just run `make help`.

After reviewing the possible targets it's quick and easy to get bpfd deployed locally on your system via a [KIND cluster](https://kind.sigs.k8s.io/). with:

```bash
make run-on-kind
```

> **NOTE**

Reviewer: ```suggestion
**NOTE**

```
Not sure if adding a new line does it, but **NOTE** is on the same line as the next sentence when the markdown is rendered. Otherwise, add a `:` to NOTE and leave it on the same line.

Author: Removed it :)

The container images used for `bpfd`,`bpfd-agent`, and `bpfd-operator` can also be

> configured, by defualt local image builds will be used for the kind deployment.

Reviewer: ```suggestion
configured, by default local image builds will be used for the kind deployment.
```

Author: Fixed

    creationTimestamp: "2022-12-07T22:41:29Z"
    finalizers:
    - bpfd.io.agent/finalizer
    generation: 2
    labels:
      owningConfig: xdp-pass-all-nodes
    name: xdp-pass-all-nodes-bpfd-deployment-worker2
    ownerReferences:
    - apiVersion: bpfd.io/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: EbpfProgramConfig
      name: xdp-pass-all-nodes
      uid: 6e3f5851-97b1-4772-906b-3ac69c6a4057
    resourceVersion: "1506"
    uid: 384d3d5c-e62b-4be3-9bf0-c6cf0e315acf
  spec:
    programs:
      bdeac6d3-4128-464e-9161-6010684eca27:
        attachpoint:
          interface: eth0
        maps: {}
  status:
    conditions:
    - lastTransitionTime: "2022-12-07T22:41:30Z"
      message: Successfully loaded ebpfProgram
      reason: bpfdLoaded
      status: "True"
      type: Loaded
```

Applications wishing to use bpfd to deploy/manage their bpf programs in kubernetes will make use of this
object to find references to the bpfMap pin points (`spec.maps`) in order to configure their bpf programs.

### Controllers

The Bpfd-Operator performs a few major functions and houses two major controllers the `bpfd-agent` and `bpfd-operator`.

#### bpf-agent

> The bpfd-agent controller is deployed alongside bpfd in a daemonset.  It's main purpose to to watch user intent (in ebpfProgramConfig Objects) and communicate with

Reviewer: ```suggestion
The bpfd-agent controller is deployed alongside bpfd in a daemonset.  It's main purposeis to watch user intent (in ebpfProgramConfig Objects) and communicate with
```

> bpfd via an mTLS secured connection in order to translate the cluster-wide user-inetent to per node state.

Reviewer: ```suggestion
bpfd via a mTLS secured connection in order to translate the cluster-wide user-intent to per node state.
```
Not sure if it's `a` or `an`, I think `a`.

Author: Done
